cryptography/src/bls12381/primitives/ops.rs

```diff
@@ -99,53 +192,95 @@ pub fn partial_aggregate(
     poly::Signature::recover(threshold, partials)
 }
 
-/// Aggregates multiple signatures over unique messages from the same public key.
+/// Aggregates multiple public keys.
+///
+/// # Warning
 ///
-/// If the same signatures is provided multiple times, the function will not error
-/// but any attempt to verify the aggregated signature will fail.
-pub fn aggregate(signatures: &[group::Signature]) -> group::Signature {
+/// This function assumes a group check was already performed on each `public`.
+pub fn aggregate_public_keys(public_keys: &[group::Public]) -> group::Public {
+    let mut p = group::Public::zero();
+    for pk in public_keys {
+        p.add(pk);
+    }
+    p
+}
+
+/// Aggregates multiple signatures.
+///
+/// # Warning
+///
+/// This function assumes a group check was already performed on each `signature`.
+pub fn aggregate_signatures(signatures: &[group::Signature]) -> group::Signature {
     let mut s = group::Signature::zero();
     for sig in signatures {
         s.add(sig);
     }
     s
 }
 
-/// Verifies the aggregate signature over multiple unique messages from the same public key.
+/// Verifies the aggregate signature over a single message from multiple public keys.
+///
+/// # Warning
 ///
-/// If the same message is provided multiple times, the function will error.
-pub fn verify_aggregate(
+/// This function assumes the caller has performed a group check and collected a proof-of-possession
+/// for all provided `public`. This function assumes a group check was already performed on the
+/// `signature`. It is not safe to provide duplicate public keys.
+pub fn aggregate_verify_multiple_public_keys(
+    public: &[group::Public],
+    namespace: Option<&[u8]>,
+    message: &[u8],
+    signature: &group::Signature,
+) -> Result<(), Error> {
+    // Aggregate public keys
+    //
+    // We can take advantage of the bilinearity property of pairings to aggregate public keys
+    // that have all signed the same message (as long as all public keys are unique).
+    let agg_public = aggregate_public_keys(public);
+
+    // Verify the signature
+    verify_message(&agg_public, namespace, message, signature)
+}
+
+/// Verifies the aggregate signature over multiple unique messages from a single public key.
+///
+/// # Warning
+///
+/// This function assumes a group check was already performed on `public` and `signature`. It is not
+/// safe to provide an aggregate public key or to provide duplicate messages.
+///
+/// ## Why not aggregate public keys?
```

Comment:
FWIW there is an optimization that can be done (reduces number of pairings checks) for multiple public keys signing multiple messages but figured this PR was both big enough/that use case isn't as common.